// js-excel-template/nodejs/common.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
class JsExcelTemplateBase {
    constructor(workbook) {
        this.workbook = workbook;
    }
    set(name, value) {
        if (Array.isArray(value)) {
            if (value.length === 0) {
                return;
            }
            for (const sheetName of this.workbook.SheetNames) {
                const sheet = this.workbook.Sheets[sheetName];
                const targetRowIndex = this.findRowIndex(name, sheet);
                if (targetRowIndex >= 0) {
                    const ref = sheet["!ref"];
                    if (ref) {
                        const [leftRange, rightRange] = ref.split(":");
                        const { column: maxColumnName, rowIndex: maxRowCount } = this.parseCellName(rightRange);
                        const newMaxRowCount = maxRowCount + value.length;
                        if (maxRowCount >= 0) {
                            sheet["!ref"] = `${leftRange}:${maxColumnName}${newMaxRowCount}`;
                        }
                    }
                    for (const cellName in sheet) {
                        if (sheet.hasOwnProperty(cellName) && cellName.indexOf("!") !== 0) {
                            const { column, rowIndex } = this.parseCellName(cellName);
                            const cell = sheet[cellName];
                            if (rowIndex === targetRowIndex && cell.w) {
                                const index = cell.w.indexOf(`{${name}.`);
                                if (index >= 0) {
                                    let fieldName;
                                    for (let i = index + `{${name}.`.length; i < cell.w.length; i++) {
                                        if (cell.w[i] === "}") {
                                            fieldName = cell.w.substring(index + `{${name}.`.length, i);
                                            break;
                                        }
                                    }
                                    if (fieldName) {
                                        for (let i = 1; i < value.length; i++) {
                                            sheet[column + (rowIndex + i)] = JSON.parse(JSON.stringify(sheet[cellName]));
                                        }
                                        for (let i = 0; i < value.length; i++) {
                                            const newCell = sheet[column + (rowIndex + i)];
                                            this.setCell(newCell, `{${name}.${fieldName}}`, value[i][fieldName]);
                                        }
                                    }
                                }
                            }
                        }
                    }
                    return;
                }
            }
        }
        else {
            for (const sheetName of this.workbook.SheetNames) {
                const sheet = this.workbook.Sheets[sheetName];
                for (const cellName in sheet) {
                    if (sheet.hasOwnProperty(cellName) && cellName.indexOf("!") !== 0) {
                        const cell = sheet[cellName];
                        this.setCell(cell, `{${name}}`, value);
                    }
                }
            }
        }
    }
    setCell(cell, name, value) {
        if (cell.v && typeof cell.v === "string" && cell.v.indexOf(name) >= 0) {
            cell.v = cell.v === name ? value : cell.v.split(name).join(value);
            if (typeof cell.v === "number") {
                cell.t = "n";
            }
            else if (Object.prototype.toString.call(cell.v) === "[object Date]") {
                cell.t = "d";
            }
            else if (cell.v === true || cell.v === false) {
                cell.t = "b";
            }
        }
    }
    parseCellName(cellName) {
        for (let i = 0; i < cellName.length; i++) {
            if (!isNaN(+cellName[i])) {
                return {
                    column: cellName.substring(0, i),
                    rowIndex: +cellName.substring(i),
                };
            }
        }
        return {
            column: "",
            rowIndex: -1,
        };
    }
    findRowIndex(name, sheet) {
        for (const cellName in sheet) {
            if (sheet.hasOwnProperty(cellName) && cellName.indexOf("!") !== 0) {
                const cell = sheet[cellName];
                if (cell.w && cell.w.indexOf(`{${name}.`) >= 0) {
                    const { rowIndex } = this.parseCellName(cellName);
                    return rowIndex;
                }
            }
        }
        return -1;
    }
}
exports.default = JsExcelTemplateBase;
